### PR TITLE
String pattern alias

### LINKED ISSUE #2294: ✨ Allow assigning names to string prefix patterns.
Gleam allows you to assign names to subpatterns for usage in the expression body using `as`. This is useful for many reasons but one reason is to deduplicate work in combination with alternative clause patterns.

We could write something like:
```gleam
case list {
  [1 as x, ..rest] | [2 as x, ..rest] -> 
    // do something with `x` here

  ... ->
    ...
}
```

Unfortunately, the equivalent for string patterns is a syntax error:
```gleam
case string {
  "a" as x <> rest | "b" as x <> rest ->
    // do something with `x` here

  ... ->
    ...
}
```

Instead, you'd have to write something like:
```gleam
case string {
  "a" <> rest | "b" <> rest -> {
    let assert Ok(#(x, rest)) = string.pop_grapheme(string)
    // do something with `x`
  }

  ... ->
    ...
}
```

Which both duplicates work we've already done and makes the code less immediately clear what it does.

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This should close #2294 and #2243

### CODE REVIEW

**Comment by lpil** [comment]
Hi @giacomocavalieri ! Would you like me to pick this work?

**Comment by giacomocavalieri** [comment]
@lpil sorry if it took long, I think the PR is ready for review now 👍

File: compiler-core/src/type_.rs [hunk 1]
```diff
@@ -676,6 +693,7 @@ impl ValueConstructor {
     pub(crate) fn get_documentation(&self) -> Option<&str> {
         match &self.variant {
             ValueConstructorVariant::LocalVariable { .. } => Some("A locally defined variable."),
+            ValueConstructorVariant::LocalConstant { .. } => Some("A locally defined constant."),
```
Comment (lpil) [inline compiler-core/src/type_.rs:696]:
```suggestion
            ValueConstructorVariant::LocalConstant { .. } => Some("A locally defined variable."),
```

This is an implementation detail of the compiler, it's still a regular variable from the user's point of view.

File: compiler-core/src/type_/pattern.rs [hunk 2]
```diff
@@ -75,6 +75,47 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
...
+                        let initial_typ = initial.type_.clone();
+                        unify(initial_typ, literal.type_())
+                    }
+
+                    // This variable was not defined in the Initial multi-pattern
+                    _ => Err(UnifyError::ExtraVarInAlternativePattern { name: name.into() }),
+                }
+            }
+        }
+    }
```
Comment (lpil) [inline compiler-core/src/type_/pattern.rs:117]: We only ever insert strings so the vast majority of the code head is dead code and cannot be tested in any way. We can instead have a `insert_constant_string` function that does much less, and we can know that it works.

**Review by lpil — COMMENTED** [review]
Fab! Two tiny little things inline.

**Review by lpil — APPROVED** [review]

**Pull request merged**